### PR TITLE
pdngen: speed up PDN insertion

### PR DESCRIPTION
**Pull request by oharboe**
@maliberty @gadfort A stab in the dark... Runs MUCH faster, but does it still do the right thing?

Signed-off-by: Øyvind Harboe <oyvind.harboe@zylin.com>

### CODE REVIEW

File: src/pdn/src/shape.cpp [hunk 1]
```diff
@@ -627,17 +627,7 @@ const odb::Rect FollowPinShape::getMinimumRect() const
 bool FollowPinShape::cut(const ShapeTree& obstructions,
```
Comment (QuantamHD) [inline src/pdn/src/shape.cpp:627]: I would personally just delete this function now. It's not providing any value. Calling Shape::cut(filter_obstructions, replacements) is probably better.

**Comment by QuantamHD** [comment]
Looks like some minor issues here and there. This does look the same, but there are small differences

In the region_temp_sensor test two lines are swapped in the DEF

```
Differences found at line 80.
    - via5_6_1600_1600_1_1_1600_1600 + VIARULE M4M5_PR + CUTSIZE 800 800  + LAYERS met4 via4 met5  + CUTSPACING 800 800  + ENCLOSURE 400 190 310 400  ;
    - via5_6_1600_5000_3_1_1600_1600 + VIARULE M4M5_PR + CUTSIZE 800 800  + LAYERS met4 via4 met5  + CUTSPACING 800 800  + ENCLOSURE 400 190 310 500  + ROWCOL 3 1  ;
```

In the region_secondary_nets it's missing a stripe, but looks identical otherwise
```
NEW met4 1600 + SHAPE STRIPE ( 156420 63500 ) ( 156420 186500 )
```

File: src/pdn/src/shape.cpp (hunk 1 above)
Comment (oharboe) [reply src/pdn/src/shape.cpp:627]: Makes sense, but the build failures need to be investigated first. Some of those are irrelevant, docker failures, but others appear relevant.

I don't know the details of what is going on here...

```
[deleted]
power_switch_regions (tcl) *FAIL*
[deleted]
region_temp_sensor (tcl) *FAIL*
region_secondary_nets (tcl) *FAIL*
[deleted]
```


```
*** /tmp/workspace/OpenROAD-Public_PR-2795-head/src/pdn/test/region_temp_sensor.ok	2023-01-20 06:20:47.589640738 +0000
--- /tmp/workspace/OpenROAD-Public_PR-2795-head/src/pdn/test/results/region_temp_sensor-tcl.log	2023-01-20 06:27:04.138421493 +0000
***************
*** 17,20 ****
  [INFO ODB-0133]     Created 155 nets and 491 connections.
  [INFO PDN-0001] Inserting grid: Core
  [INFO PDN-0001] Inserting grid: TempSensor
! No differences found.
--- 17,22 ----
  [INFO ODB-0133]     Created 155 nets and 491 connections.
  [INFO PDN-0001] Inserting grid: Core
  [INFO PDN-0001] Inserting grid: TempSensor
! Differences found at line 80.
!     - via5_6_1600_1600_1_1_1600_1600 + VIARULE M4M5_PR + CUTSIZE 800 800  + LAYERS met4 via4 met5  + CUTSPACING 800 800  + ENCLOSURE 400 190 310 400  ;
!     - via5_6_1600_5000_3_1_1600_1600 + VIARULE M4M5_PR + CUTSIZE 800 800  + LAYERS met4 via4 met5  + CUTSPACING 800 800  + ENCLOSURE 400 190 310 500  + ROWCOL 3 1  ;
*** /tmp/workspace/OpenROAD-Public_PR-2795-head/src/pdn/test/region_secondary_nets.ok	2023-01-20 06:20:47.589640738 +0000
--- /tmp/workspace/OpenROAD-Public_PR-2795-head/src/pdn/test/results/region_secondary_nets-tcl.log	2023-01-20 06:27:04.319437740 +0000
***************
*** 26,29 ****
  [INFO ODB-0133]     Created 260 nets and 856 connections.
  [INFO PDN-0001] Inserting grid: Core
  [INFO PDN-0001] Inserting grid: Region
! No differences found.
--- 26,31 ----
  [INFO ODB-0133]     Created 260 nets and 856 connections.
  [INFO PDN-0001] Inserting grid: Core
  [INFO PDN-0001] Inserting grid: Region
! Differences found at line 723.
!       NEW met4 1600 + SHAPE STRIPE ( 156420 63500 ) ( 156420 186500 )
!       NEW met4 1600 + SHAPE STRIPE ( 146420 63500 ) ( 146420 186500 )
*** /tmp/workspace/OpenROAD-Public_PR-2795-head/src/pdn/test/power_switch_regions.ok	2023-01-20 06:20:47.585640374 +0000
--- /tmp/workspace/OpenROAD-Public_PR-2795-head/src/pdn/test/results/power_switch_regions-tcl.log	2023-01-20 06:27:16.881565274 +0000
***************
*** 20,23 ****
  [INFO ODB-0133]     Created 157 nets and 491 connections.
  [INFO PDN-0001] Inserting grid: Core
  [INFO PDN-0001] Inserting grid: TempSensor
! No differences found.
--- 20,25 ----
  [INFO ODB-0133]     Created 157 nets and 491 connections.
  [INFO PDN-0001] Inserting grid: Core
  [INFO PDN-0001] Inserting grid: TempSensor
! Differences found at line 71.
! VIAS 13 ;
! VIAS 12 ;
```

**Comment by oharboe** [comment]
> Looks like some minor issues here and there. This does look the same, but there are small differences

I tried to modify the code so that it would have no differences, but I don't know the code at all, so I presume there are some side-effects that my change is susceptible to?

Why would there be any differences at all?

**Comment by gadfort** [comment]
It's failing because you cannot apply the filter for all grid components (that filtering only applies to followpins). What you have breaks anything that have multiple grids defined (like with regions). I think I see a possible solution, but I can't get to it till this weekend.

**Comment by oharboe** [comment]
> It's failing because you cannot apply the filter for all grid components (that filtering only applies to followpins). What you have breaks anything that have multiple grids defined (like with regions). I think I see a possible solution, but I can't get to it till this weekend.

Great! The value of this pull request was to verify the location of the performance issue, it was too much to hope that it would be so simple to fix it.

**Comment by maliberty** [comment]
replaced by https://github.com/The-OpenROAD-Project/OpenROAD/pull/2803

**Pull request closed without merging**